### ISSUE TITLE
skew_normal_rng overflowing

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/math#527]
#### Summary:

This program

```
generated quantities {
  real y = skew_normal_rng(1, 10, -4);
}
```

generates `-inf` draws for `y`, but it shouldn't.

#### Reproducible Steps:

See above.

#### Current Output:

From CmdStan:

```
                Mean     MCSE   StdDev   5%   50%  95%  N_Eff  N_Eff/s    R_hat
lp__            0.00  0.0e+00  0.0e+00 0.00  0.00 0.00   1000    37997      nan
accept_stat__   0.00  0.0e+00  0.0e+00 0.00  0.00 0.00   1000    37997      nan
y                inf      nan      nan  -19  -5.6  2.0   1000    37997      nan
```

#### Expected Output:

Should have finite posterior mean for `y`.

#### Additional Information:

If you code `y` as a parameter and use a sampling statement in the model, this problem doesn't arise, but there are a ton of divergent transitions.  So maybe we also need to suggest an alternative parameterization of skew normal?

#### Current Version:
v2.15.0

**Comment by sakrejda** [comment]
This sounds like one easy issue (look up the right RNG and see where our
generator goes wrong) plus one potentially complicated issue (divergences,
which could just be parameterization or it could be the math functions not
being accurate)

**Comment by bgoodri** [comment]
I already looked it up. We are just inputing a random standard uniform
variate into the quantile function from Boost's implementation of the skew
normal distribution. But apparently Boost's implementation involves
numerical root-finding.

**Comment by sakrejda** [comment]
There's an rng for the skew normal that involves 2 X ~ N(0,1) draws and a
short calculation. I'll put a link in here and leave it as a new dev issue
when I find the procedure again.

**Comment by bgoodri** [comment]
Yeah, that paper is by Mudholkar and Hutson.

**Comment by sakrejda** [comment]
Here's the thing I ran into---didn't check if they're matching
parameterizations: http://azzalini.stat.unipd.it/SN/faq-r

**Comment by bbbales2** [comment]
I gotta fix this to get the vectorization tests working nicely.

It looks like there are two options, both from Azzalini (both get rid of the infs I was seeing). This is what he uses in his R package (sn):

(mu is location, sigma is scale, alpha is shape)

```c++
double r1 = normal_rng(0.0, 1.0, rng),
  r2 = normal_rng(0.0, 1.0, rng);

if(r2 > alpha * r1)
  r1 *= -1;

output[n] = mu + sigma * r1;
```

This is one of the solutions he gives in his paper (https://arxiv.org/pdf/0911.2342.pdf):

```c++
double delta = alpha / sqrt(1 + alpha * alpha);
double r1 = normal_rng(0.0, 1.0, rng),
  r2 = normal_rng(0.0, 1.0, rng);

double r3 = std::abs(r1) * delta + r2 * sqrt(1 - delta * delta);

output[n] = mu + sigma * r3;
```

The first is way simpler, but it comes from GPL3 code so I suppose I can't just copy it. Is it obvious how to get to the first thing from the second? If not I'll just use the second. This isn't performance sensitive stuff anyway.

**Comment by sakrejda** [comment]
The first one is in the paper too: it's a simulation from a standard bivariate normal with a correlation calculated from the alpha parameter (around equation 8) and then you take the larger of the two.

**Comment by bob-carpenter** [comment]
Algorithms aren't subject to copyright.  So you can code the algorithm in the paper and it won't fall under copyleft from another implementation.

**Comment by sakrejda** [comment]
That's the point, no need to lift it from the code.

**Comment by rok-cesnovar** [comment]
This is not an issue on develop. I guess it was fixed in #622.
